# Amplicon placement

### Run the install script
```
./amplicon_installUbuntu.sh
```

## Generate amplicon VCF file

**Make sure you have these files in the current directory (`usher/scripts/amplicon`)**
- For faToVcf to mask problematic reference sites: `problematic_sites_sarsCov2.vcf`
- SARS-CoV-2 reference genome: `wuhCor1.fa`
- Your amplicon reads fastq file

<br>

### Run the script like this, passing in your amplicon reads fastq file as the only argument:

```
./generate_vcf.sh <amplicon_reads.fastq>
```

This should generate a vcf file in the current directory called `amplicons.vcf`

## Placing amplicon sequences using UShER
<br>

Navagate back to top level `usher` directory. Place the amplicon sequences by running the following command with input arguments:
- `-i`: MAT protobuf to place amplicon samples onto
- `-v`: the generated vcf file from the previous steps
- `-f`: an aligned multifasta file generated by `generate_vcf.sh` script containing amplicon sample to reference alignment start and end coordinate information that is necessary for placement using UShER 
- `-d`: output directory, defaults to current directory

Example usage:
```
build/amplicon -i public-latest.all.masked.pb -v scripts/amplicon/output.vcf -f scripts/amplicon/amplicon.fa -d output/
```