# Amplicon placement

## Install UShER 
- Build from source and run the correct install script for your OS

```
git clone https://github.com/mrkylesmith/usher.git
cd usher
./install/installUbuntu_amplicon.sh
cd scripts/amplicon
```

COMING SOON:
- Run inside Docker
- Conda install UShER

## Generate amplicon variant call file

- Download SARS-CoV2 reference
```
wget https://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/bigZips/wuhCor1.fa.gz
gunzip wuhCor1.fa.gz
```

- Download sites to mask 
```
wget https://raw.githubusercontent.com/W-L/ProblematicSites_SARS-CoV2/master/problematic_sites_sarsCov2.vcf
```
- Download latest public tree to place samples on
(small tree given here for testing)
```
wget http://hgdownload.soe.ucsc.edu/goldenPath/wuhCor1/UShER_SARS-CoV-2/2021/05/25/public-2021-05-25.all.masked.pb.gz
gunzip public-2021-05-25.all.masked.pb.gz
```

- Make sure your amplicon reads file (FASTQ) is in current directory (`scripts/amplicon`)

- Run `get_variants.py` script below to generate VCF for amplicon reads (`data/amplicons.vcf`) and an MSA FASTA file (`data/amplicons.fa`) which is used to obtain amplicon read alignment coordinate positions.
```
python3 get_variants.py -f <amplicon_reads.fastq> [options]
```
Note: `-f` is a required flag to give amplicon reads FASTQ file

Options:
- `-h`|`--help`: Get usage information.
- `-a`|`--align_with`: Specify to use alternate aligner. Currently supports bwa-mem2 as alternate aligner (`-a bwamem2`). Default alignment with `minimap2` if no flag is given.
- `-r`|`--remove`: Remove duplicate amplicon reads in given FASTQ file 

Once this script is run successfully, `data/amplicons.fa` and `data/amplicons.vcf` will be generated and needed for the next placement command.


## Placing amplicon sequences using UShER
<br>
Now use UShER to place the amplicon sequences on the tree previously downloaded, using the following command:

```
amplicon -i <tree.pb> -f data/amplicons.fa -v data/amplicons.vcf -d output
```

Required Options:
- `-i`: MAT protobuf to place amplicon samples onto
- `-v`: the generated vcf file from the previous steps
- `-f`: MSA FASTA file generated by the previous step that contains amplicon coordinate information necessary for placement
- `-d`: (Optional) output directory, defaults to current directory (if flag not given, files will be output to current directory)         

### Output files
The following two output files will be generated after placement and written to the output directory or current directory if `-d` flag not given.
- `final-tree.nh`:  A generated Newick file of the tree 
- `clade0.txt` and/or `clade1.txt`: Information about likely clade assignments for each amplicon sample, relative to the number of equally parsimonious placements possible for the given sample.  Samples for `Nextclade` and `pangolin` are split into separate clade.txt files. 
 
